File: .claude/scripts/user_utils.ts
#!/usr/bin/env ts-node
/**
 * User utility functions for authentication and data access.
 *
 * TypeScript implementation for consistency with Node.js/TypeScript ecosystem.
 * Uses standard npm packages: sqlite3 and bcrypt.
 */

import * as sqlite3 from 'sqlite3';
import * as bcrypt from 'bcrypt';
import * as path from 'path';

// Database path - resolved relative to script location for consistent execution
const DB_PATH = path.join(__dirname, 'users.db');

interface User {
    id: number;
    username: string;
    password_hash: string;
    email: string;
}

interface SafeUser {
    id: number;
    username: string;
    email: string;
}

/**
 * Fetch user from database by username.
 */
export async function getUserByName(username: string): Promise<User | null> {
    return new Promise((resolve, reject) => {
        const db = new sqlite3.Database(DB_PATH);

        // Use parameterized query to prevent SQL injection
        // Explicitly select required columns for deterministic ordering
        db.get(
            "SELECT id, username, password_hash, email FROM users WHERE username = ?",
            [username],
            (err, row: User | undefined) => {
                db.close();
                if (err) {
                    reject(err);
                } else {
                    resolve(row || null);
                }
            }
        );
    });
}

/**
 * Authenticate a user with username and password.
 */
export async function authenticateUser(username: string, password: string): Promise<boolean> {
    try {
        const user = await getUserByName(username);

        if (user) {
            // Use bcrypt for secure password verification
            return await bcrypt.compare(password, user.password_hash);
        }

        return false;
    } catch (error) {
        console.error('Authentication error:', error);
        return false;
    }
}

/**
 * Update user's email address.
 */
export async function updateUserEmail(userId: number, newEmail: string): Promise<boolean> {
    return new Promise((resolve, reject) => {
        const db = new sqlite3.Database(DB_PATH);

        // Use parameterized query to prevent SQL injection
        db.run(
            "UPDATE users SET email = ? WHERE id = ?",
            [newEmail, userId],
            (err) => {
                db.close();
                if (err) {
                    reject(err);
                } else {
                    resolve(true);
                }
            }
        );
    });
}

/**
 * Delete a user from the database.
 */
export async function deleteUser(userId: number): Promise<void> {
    return new Promise((resolve, reject) => {
        const db = new sqlite3.Database(DB_PATH);

        // Use parameterized query to prevent SQL injection
        db.run(
            "DELETE FROM users WHERE id = ?",
            [userId],
            (err) => {
                db.close();
                if (err) {
                    reject(err);
                } else {
                    resolve();
                }
            }
        );
    });
}

/**
 * Get all users from database.
 * SECURITY: Only returns safe fields (id, username, email) - password hashes are excluded.
 */
export async function getAllUsers(): Promise<SafeUser[]> {
    return new Promise((resolve, reject) => {
        const db = new sqlite3.Database(DB_PATH);

        // Security fix: Select only safe fields, excluding password_hash
        db.all(
            "SELECT id, username, email FROM users",
            [],
            (err, rows: SafeUser[]) => {
                db.close();
                if (err) {
                    reject(err);
                } else {
                    resolve(rows || []);
                }
            }
        );
    });
}

// CLI interface for direct script execution
if (require.main === module) {
    const command = process.argv[2];
    const args = process.argv.slice(3);

    (async () => {
        try {
            switch (command) {
                case 'list':
                    const users = await getAllUsers();
                    console.log('Users:', users);
                    break;
                case 'auth':
                    if (args.length < 2) {
                        console.error('Usage: ts-node user_utils.ts auth <username> <password>');
                        process.exit(1);
                    }
                    const isAuthenticated = await authenticateUser(args[0], args[1]);
                    console.log('Authenticated:', isAuthenticated);
                    break;
                case 'update-email':
                    if (args.length < 2) {
                        console.error('Usage: ts-node user_utils.ts update-email <user_id> <new_email>');
                        process.exit(1);
                    }
                    await updateUserEmail(parseInt(args[0]), args[1]);
                    console.log('Email updated successfully');
                    break;
                case 'delete':
                    if (args.length < 1) {
                        console.error('Usage: ts-node user_utils.ts delete <user_id>');
                        process.exit(1);
                    }
                    await deleteUser(parseInt(args[0]));
                    console.log('User deleted successfully');
                    break;
                default:
                    console.error('Unknown command:', command);
                    console.error('Available commands: list, auth, update-email, delete');
                    process.exit(1);
            }
        } catch (error) {
            console.error('Error:', error);
            process.exit(1);
        }
    })();
}

File: .claude/scripts/user_utils.py
#!/usr/bin/env python3
"""
User utility functions for authentication and data access.

Note: This is a Python utility script in a Node.js/TypeScript project.
It is intended for AI agent internal sandbox use only. For production use,
consider implementing these utilities in TypeScript/Node.js to align with
the project's primary language and dependency ecosystem.
"""

import sqlite3
import os
import bcrypt


# Database path - resolved relative to script location for consistent execution
DB_PATH = os.path.join(os.path.dirname(__file__), "users.db")


def get_user_by_name(username):
    """Fetch user from database by username."""
    with sqlite3.connect(DB_PATH) as conn:
        # Use Row factory to allow access by column name
        conn.row_factory = sqlite3.Row
        cursor = conn.cursor()

        # Use parameterized query to prevent SQL injection
        # Explicitly select required columns for deterministic ordering
        cursor.execute("SELECT id, username, password_hash, email FROM users WHERE username = ?", (username,))

        result = cursor.fetchone()
        return result


def authenticate_user(username, password):
    """Authenticate a user with username and password."""
    user = get_user_by_name(username)

    if user:
        # Access by column name instead of brittle index
        stored_hash = user['password_hash']

        # Use bcrypt for secure password verification
        if bcrypt.checkpw(password.encode(), stored_hash.encode() if isinstance(stored_hash, str) else stored_hash):
            return True

    return False


def update_user_email(user_id, new_email):
    """Update user's email address."""
    with sqlite3.connect(DB_PATH) as conn:
        cursor = conn.cursor()

        # Use parameterized query to prevent SQL injection
        cursor.execute("UPDATE users SET email = ? WHERE id = ?", (new_email, user_id))
        conn.commit()

    return True


def delete_user(user_id):
    """Delete a user from the database."""
    with sqlite3.connect(DB_PATH) as conn:
        cursor = conn.cursor()
        # Use parameterized query to prevent SQL injection
        cursor.execute("DELETE FROM users WHERE id = ?", (user_id,))
        conn.commit()


def get_all_users():
    """Get all users from database."""
    with sqlite3.connect(DB_PATH) as conn:
        cursor = conn.cursor()
        cursor.execute("SELECT id, username, email FROM users")
        users = cursor.fetchall()
        return users
